### PR TITLE
Version 2

### LINKED ISSUE #1: Tree.onSelect
It might be cleaner to have an onSelect callback that can be placed on the root tree node. It only get's fired when an item gets selected. A group can only be selected during multi select.

New Api:

```ts
type TreeProps = {
  onSelect: () // only get's called when a single item is selected.
}
```

Then when the tree selects something programmatically, this code can run and does not need to be duplicated.

### LINKED ISSUE #6: Re-think handlers.select option.
`handlers.select(e, {selectOnClick: true})`

That's weird.

### LINKED ISSUE #8: Make a Default Node Component.
This will help users get up and running quickly, without needing to create their own styles at first. @mason-fish had this idea.

### LINKED ISSUE #9: Add keyboard interaction and ARIA roles
I saw your post on reddit.  Nice tree control.

I wanted to give a suggestion that could help take it to the next level.  Right now it is very mouse oriented.  However, not everyone may prefer to use the mouse.  If you look up the the [tree role on MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/tree_role#keyboard_interactions) it documents keyboard interactions for a tree.  That way users can operate it without taking their hands off the keyboard if they prefer (similar to Ctrl-C Ctrl-V for copy and paste).

### LINKED ISSUE #30: How to get the current scroll position and the overall height of the scroll container?
I want to make a tree which would load data asynchronously upon user opening a tree node (which seems easy to do here) or upon scrolling down the tree root (in my use case root would have not only folders, but also lots and lots of items (thoughsands of them) and it wouldn't be a great experience for a user to load them all at once).
Is there a way to get the overall height of the scroll container and current scroll position to make the latter possible? Or if not, is there a better way to do what I need?

### LINKED ISSUE #38: Drag Handle
Love what you have made!!

Learning so much by reading the source. Fantastic job.

I would love a drag handle ref option. 

packages/react-arborist/src/components/row.tsx   maybe expose optionally the dragRef ?

Thanks!

### LINKED ISSUE #41: Dragging parent into a child - critical issue
When a parent folder is being dragged into a child there is no check to prohibit this from happening.
This causes the entire subtree to disappear.

### LINKED ISSUE #44: Expose onSelect event to Tree component
Hi James,

I need to listen for the tree node select event via the Tree component.
Would it be possible to add this?

My use case is I'm using check-boxes to select multiple tree nodes using the `handlers.select` which has shift-click but I only want one event to fire.

Kind regards,
Marten

### LINKED ISSUE #46: Error with multiple trees
If you have 2 Tree components rendering at the same time a React `Maximum update depth exceeded` error is thrown. The line below is causing the error.

```js
useLayoutEffect(() => {
    // @ts-ignore
    dispatch(actions.setVisibleIds(api.visibleIds, api.idToIndex));  // <- causes the error.
  }, [dispatch, api.visibleIds, api.idToIndex, props.root]);
```

Here is a codesandbox with a basic repro.

https://codesandbox.io/s/react-arborist-forked-xb7ezj

Thanks!

### PR DESCRIPTION
**Pull request by jameskerr**
- [x] Keyboard navigation
- [x] Aria Roles fixes (fixes #9)
- [x] Smart "scrollTo" behavior
- [x] Automatically sync selection
- [x] Tree filtering now handled by arborist
- [x] Moves "openness" state into the tree with an "initialOpens" prop
- [x] Simple NodeApi class for use in the node renderer
- [x] Ability to specify the drag handler (fixes #38)
- [x] Splits the NodeRenderer into a RowRenderer and a NodeRenderer with defaults resulting in less boilerplate. (fixes #8)
- [x] Adds an onActivate callback
- [x] Adds an onSelect callback (fixes #44)
- [x] Adds an onCreate callback
- [x] Adds an onDelete callback
- [x] Adds an onScroll callback (fixes #30)
- [x] Improved abstractions for focus, selection, and activation (fixes #1, fixes #6)
- [x] Fixes the bug where you can drag a parent into itself (fixes #41)

Bugs:
- fixes #46 

# Remaining Items

- [x] Document the new APIs
- [x] Make some Demos
- [x] Fix the bugs that surface while doing the above

### CODE REVIEW

File: packages/react-arborist/src/types.ts [hunk 1]
```diff
@@ -1,148 +0,0 @@
...
-  isOpen: boolean;
-  isSelected: boolean;
-  isHoveringOverChild: boolean;
-  isDragging: boolean;
-  isSelectedStart: boolean;
-  isSelectedEnd: boolean;
-  isEditing: boolean;
-};
-
-export type NodeHandlers = {
```
Comment (shorttermproject) [inline packages/react-arborist/src/types.ts:62]: When I tried this rc branch with my project, I was getting an error where the type `NodeHandlers` is missing. I saw that it was still used in the demo code for this branch and I am wondering what is the alternative type.

Comment (jameskerr) [reply packages/react-arborist/src/types.ts:62]: @shorttermproject There is no more NodeHandlers. There is now a NodeApi instance. More docs to come, but here's how it works now in some code:

https://github.com/brimdata/react-arborist/blob/version-2/packages/showcase/pages/simple-tree/index.tsx#L16-L34

Comment (jameskerr) [reply packages/react-arborist/src/types.ts:62]: @shorttermproject I've updated the README in this branch with examples and the API reference. It should be complete enough to get going.

**Review by jameskerr — COMMENTED** [review]
Approved

**Comment by mooijtech** [comment]
Hi @jameskerr, I have this running and everything works (with checkboxes). Thank you!

https://user-images.githubusercontent.com/64754391/197047335-d65a7351-099d-4a1b-8911-99e49d020394.mp4

**Comment by jameskerr** [comment]
@mooijtech YES! Thanks for posting that.

**Pull request merged**